docs/lang/articles/reference.md

```diff
@@ -0,0 +1,172 @@
+about your program behavior, then you are coming to the correct place. If you
+find the actual behavior different from the one described in this article, feel
+free to create an [issue](https://github.com/taichi-dev/taichi/issues/new/choose).
+Anything unspecified in this article is subject to change, so you should not
+rely on it in your programs.
+
+**To contributors**: This article specifies what the language *should* be. That
+is, you should try to match the implementation of the Taichi compiler with this
+article. You can clearly determine a certain behavior is *correct*, *buggy*, or
+*undefined* from this article.
+
+## Introduction
+
+Taichi is a domain-specific language embedded in Python.
+[Kernels and functions](basic/syntax.md) clearly defines the boundary between
+the Taichi language and the Python language - code in the Taichi scope is
+treated as the former, while code in the Python scope is treated as the latter.
+It should be emphasized that this article is about *the Taichi language*.
+
+That said, because Taichi is embedded in Python, the syntax of Taichi is a
+subset of that of Python. To make life easier, this article is modeled after
+the [Python language reference](https://docs.python.org/3/reference/). The
+[notation](https://docs.python.org/3/reference/introduction.html#notation) and
+[lexical analysis](https://docs.python.org/3/reference/lexical_analysis.html)
+parts exactly follow Python, so familiarize yourself with them if you are new
+to them.
+
+## Basic concepts
+
+Before detailing syntax and semantics in the next few chapters, many important
+basic concepts and general evaluation principles unique to Taichi are
+introduced here.
+
+### Values and types
+
+Like other programming languages, each expression in Taichi will be evaluated
+to a value, and each value has a type. Because Taichi provides easy interaction
+with Python and [meta-programming](advanced/meta.md) support, there are
+actually two kinds of evaluation: *compile-time evaluation* and *runtime
+evaluation*. There are also two kinds of values: *Python values* and *Taichi
+values*.
+
+:::note
+For readers familiar with programming language terms, such behavior is inspired
+by [multi-stage programming](https://en.wikipedia.org/wiki/Multi-stage_programming)
+or [partial evaluation](https://en.wikipedia.org/wiki/Partial_evaluation).
+:::
+
+A Python value is simply a
+[Python object](https://docs.python.org/3/reference/datamodel.html),
+which directly comes from the following sources:
+- Free variables
+- Arguments passed via `ti.template()`
+- Literals
+
+Furthermore, as long as all the operands of an operation are Python values,
+compile-time evaluation will take place and a result Python value will be
+produced. For meta-programming purposes, Taichi provides an advanced
+environment for compile-time evaluation: `ti.static()`, where more operations
+are supported.
+
+A Python value only exists at compile time. After compile-time evaluation, all
```

Comment:
I feel like it's more like "transformed into Taichi values at JIT compile time"?

Reply:
I understand your point. The accurate way to say this should be "transformed into literals of Taichi AST". However, introducing Taichi AST here may complicate things even further. My purpose is to let users directly think about the final execution behavior based on their source code, and the actual Taichi value is indeed created at runtime.